File: common/src/main/java/juuxel/adorn/block/TableBlock.java
package juuxel.adorn.block;

import juuxel.adorn.util.FourWayShapeConfig;
import net.minecraft.world.level.block.state.BlockBehaviour.Properties;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.resources.Identifier;
import net.minecraft.core.Direction;
import net.minecraft.world.phys.shapes.VoxelShape;
import org.jetbrains.annotations.Nullable;

public final class TableBlock extends AbstractTableBlock implements BlockWithDescription {
    private static final String DESCRIPTION_KEY = "block.adorn.table.description";
    private static final VoxelShape[] SHAPES = new VoxelShape[32];

    static {
        var topShape = box(0.0, 14.0, 0.0, 16.0, 16.0, 16.0);
        var legX0Z0 = box(1.0, 0.0, 1.0, 4.0, 14.0, 4.0);
        var legX1Z0 = box(12.0, 0.0, 1.0, 15.0, 14.0, 4.0);
        var legX0Z1 = box(1.0, 0.0, 12.0, 4.0, 14.0, 15.0);
        var legX1Z1 = box(12.0, 0.0, 12.0, 15.0, 14.0, 15.0);
        FourWayShapeConfig nonCarpetedConfig = new FourWayShapeConfig(topShape, legX0Z0, legX1Z0, legX0Z1, legX1Z1);
        FourWayShapeConfig carpetedConfig = nonCarpetedConfig.addToBaseShape(CARPET_SHAPE);
        var booleans = new boolean[] { true, false };

        for (var north : booleans) {
            for (var east : booleans) {
                for (var south : booleans) {
                    for (var west : booleans) {
                        for (var hasCarpet : booleans) {
                            var key = getShapeKey(north, east, south, west, hasCarpet);
                            var config = hasCarpet ? carpetedConfig : nonCarpetedConfig;
                            var shape = config.makeShape(north, east, south, west);
                            SHAPES[key] = shape;
                        }
                    }
                }
            }
        }
    }

    public TableBlock(Properties settings) {
        super(settings.forceSolidOn());
    }

    @Override
    public @Nullable Identifier getSittingStat() {
        return null;
    }

    @Override
    public String getDescriptionKey() {
        return DESCRIPTION_KEY;
    }

    @Override
    protected boolean isSittingEnabled() {
        return false;
    }

    @Override
    protected boolean canConnectTo(BlockState state, Direction sideOfSelf) {
        return state.getBlock() instanceof TableBlock;
    }

    @Override
    protected VoxelShape getShapeForKey(int key) {
        return SHAPES[key];
    }
}

File: src/main/kotlin/juuxel/adorn/block/TableBlock.kt
package juuxel.adorn.block

import it.unimi.dsi.fastutil.bytes.Byte2ObjectMap
import it.unimi.dsi.fastutil.bytes.Byte2ObjectOpenHashMap
import juuxel.adorn.api.block.BlockVariant
import juuxel.adorn.config.AdornConfigManager
import net.minecraft.block.BlockState
import net.minecraft.util.math.Direction
import net.minecraft.util.shape.VoxelShape
import net.minecraft.util.shape.VoxelShapes

open class TableBlock(variant: BlockVariant) : AbstractTableBlock(variant.createSettings()) {
    override val sittingYOffset = 0.6

    override fun isSittingEnabled() = AdornConfigManager.CONFIG.sittingOnTables

    override fun canConnectTo(state: BlockState, sideOfSelf: Direction) =
        state.block is TableBlock

    override fun getShapeForKey(key: Byte): VoxelShape = SHAPES[key]

    companion object {
        private val SHAPES: Byte2ObjectMap<VoxelShape>

        init {
            val topShape = createCuboidShape(0.0, 14.0, 0.0, 16.0, 16.0, 16.0)
            val legX0Z0 = createCuboidShape(1.0, 0.0, 1.0, 4.0, 14.0, 4.0)
            val legX1Z0 = createCuboidShape(12.0, 0.0, 1.0, 15.0, 14.0, 4.0)
            val legX0Z1 = createCuboidShape(1.0, 0.0, 12.0, 4.0, 14.0, 15.0)
            val legX1Z1 = createCuboidShape(12.0, 0.0, 12.0, 15.0, 14.0, 15.0)
            val booleans = setOf(true, false)

            fun makeShape(
                north: Boolean,
                east: Boolean,
                south: Boolean,
                west: Boolean,
                hasCarpet: Boolean
            ): VoxelShape {
                val parts = arrayListOf<VoxelShape?>(topShape)

                if (north || east || south || west) {
                    val trueCount = booleanArrayOf(north, east, south, west).count { it }

                    if (trueCount == 2) {
                        // Corners
                        parts += when {
                            north && west -> legX1Z1
                            north && east -> legX0Z1
                            south && west -> legX1Z0
                            south && east -> legX0Z0
                            else -> null // Straight lines ignored
                        }
                    } else if (trueCount == 1) {
                        // Ends
                        when {
                            north -> {
                                parts += legX0Z1
                                parts += legX1Z1
                            }

                            south -> {
                                parts += legX0Z0
                                parts += legX1Z0
                            }

                            east -> {
                                parts += legX0Z0
                                parts += legX0Z1
                            }

                            west -> {
                                parts += legX1Z0
                                parts += legX1Z1
                            }
                        }
                    }
                } else {
                    // No connections = all legs
                    parts += legX0Z0
                    parts += legX1Z0
                    parts += legX0Z1
                    parts += legX1Z1
                }

                if (hasCarpet) {
                    parts += CARPET_SHAPE
                }

                return parts.filterNotNull().reduce(VoxelShapes::union)
            }

            SHAPES = Byte2ObjectOpenHashMap(run {
                booleans.flatMap { north ->
                    booleans.flatMap { east ->
                        booleans.flatMap { south ->
                            booleans.flatMap { west ->
                                booleans.map { hasCarpet ->
                                    Bits.buildTableState(north, east, south, west, hasCarpet) to
                                            makeShape(north, east, south, west, hasCarpet)
                                }
                            }
                        }
                    }
                }.toMap()
            })
        }
    }
}
